server/quota/quota_manager.go

```diff
@@ -246,13 +247,26 @@ func (qm *QuotaManager) getGroupID(ctx context.Context) string {
 	return ""
 }
 
+func getIP(ctx context.Context) string {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return ""
+	}
+
+	vals := md.Get("X-Forwarded-For")
+
+	if len(vals) == 0 {
+		return ""
+	}
+	ips := strings.Split(vals[0], ",")
+	return ips[0]
+}
+
 func (qm *QuotaManager) getKey(ctx context.Context) string {
-	groupID := qm.getGroupID(ctx)
-	if groupID != "" {
+	if groupID := qm.getGroupID(ctx); groupID != "" {
 		return groupID
 	}
-	//TODO(lulu): find IP address
-	return ""
+	return getIP(ctx)
```

Comment:
we should check getKey hasn't returned an empty string (in Allow or wherever else) just so that we know we're not doing something funky

Reply:
Done